File: Heap/Intro.java
import java.util.*;

/* What is a Heap?

A Heap is a special tree-based data structure that follows a particular ordering rule called the Heap Property.

It is mainly used when we repeatedly need to get the:

Largest element → Max Heap
Smallest element → Min Heap

In coding interviews, you will usually use a heap through Priority Queue.

important Funtion 
1.pq.push(x);
2.pq.pop();
3.pq.top();
4.pq.empty();
5.pq.size();

*/
public class Intro {

    void minHeap(){
        PriorityQueue<Integer> pq = new PriorityQueue<>();

        // insert
        pq.add(10);
        pq.add(20);
        pq.add(50);
        pq.add(90);
        pq.add(22);

        System.out.println("MinHeap Top Element :"+pq.peek());
        pq.poll();

        System.out.println("MinHeap Top Element after poll :"+pq.peek());


    }
    void maxHeap(){
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
         pq.add(10);
        pq.add(20);
        pq.add(50);
        pq.add(90);
        pq.add(22);

        System.out.println("MaxHeap Top Element :"+pq.peek());
        pq.poll();

        System.out.println("MaxHeap Top Element After poll :"+pq.peek());
    }

    public static void main(String[] args) {
        Intro sol = new Intro();
        sol.minHeap();
        sol.maxHeap();
    }
}

File: Heap/Intro.cpp
/* What is a Heap?

A Heap is a special tree-based data structure that follows a particular ordering rule called the Heap Property.

It is mainly used when we repeatedly need to get the:

Largest element → Max Heap
Smallest element → Min Heap

In coding interviews, you will usually use a heap through Priority Queue.

important Funtion 
1.pq.push(x);
2.pq.pop();
3.pq.top();
4.pq.empty();
5.pq.size();

*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:

    // Max Heap
    void maxHeap() {

        priority_queue<int> pq;

        // Insert elements
        pq.push(10);
        pq.push(20);
        pq.push(15);
        pq.push(30);
        pq.push(40);

        cout << pq.top() << endl;  // 40

        pq.pop();                  // removes 40

        // cout << pq.top() << endl;  // 30
    }

    // Min Heap
    void minHeap() {

        priority_queue<int, vector<int>, greater<int>> pq;

        pq.push(10);
        pq.push(20);
        pq.push(15);
        pq.push(30);
        pq.push(40);

        cout << pq.top() << endl;  // 10

        pq.pop();                  // removes 10

        // cout << pq.top() << endl;  // 15
    }
};

int main() {

    Solution sol;

    sol.maxHeap();

    sol.minHeap();

    return 0;
}